packages/mobile/src/escrow/saga.ts

```diff
@@ -117,6 +217,125 @@ function* registerStandbyTransaction(context: TransactionContext, value: string,
   )
 }
 
+async function formEscrowWithdrawTxWithNoCode(
+  contractKit: ContractKit,
+  escrowWrapper: EscrowWrapper,
+  paymentId: string,
+  pepper: string,
+  msgHash: string,
+  addressIndex: number
+) {
+  const { privateKey } = generateDeterministicInviteCode(pepper, addressIndex)
+  const signature: string = (await contractKit.web3.eth.accounts.sign(msgHash, privateKey))
+    .signature
+  Logger.debug(
+    TAG + '@withdrawFromEscrowViaKomenci',
+    `Signed message hash signature is ${signature}`
+  )
+
+  const { r, s, v } = splitSignature(contractKit, signature)
+  const withdrawTx = escrowWrapper.withdraw(paymentId, v, r, s)
+  return withdrawTx
+}
+
+function* withdrawFromEscrowUsingPepper(komenciActive: boolean = false) {
+  try {
+    ValoraAnalytics.track(OnboardingEvents.escrow_redeem_start)
+    Logger.debug(TAG + '@withdrawFromEscrowViaKomenci', 'Withdrawing escrowed payment')
+    const phoneHashDetails: PhoneNumberHashDetails | undefined = yield call(
+      getUserSelfPhoneHashDetails
+    )
+
+    if (!phoneHashDetails) {
+      throw Error('Couldnt find own phone hash or pepper. Should never happen.')
+    }
+
+    const { phoneHash, pepper } = phoneHashDetails
+    const [contractKit, walletAddress, mtwAddress, feelessVerificationState]: [
+      ContractKit,
+      string,
+      string,
+      FeelessVerificationState
+    ] = yield all([
+      call(getContractKit),
+      call(getConnectedUnlockedAccount),
+      select(mtwAddressSelector),
+      select(feelessVerificationStateSelector),
+    ])
+
+    const komenciKit: KomenciKit = new KomenciKit(contractKit, walletAddress, {
+      url: KOMENCI_URL,
+      token: feelessVerificationState.komenci.sessionToken,
+    })
+
+    const escrowWrapper: EscrowWrapper = yield call([
+      contractKit.contracts,
+      contractKit.contracts.getEscrow,
+    ])
+
+    const escrowPaymentIds: string[] = yield call(
+      [escrowWrapper, escrowWrapper.getReceivedPaymentIds],
+      phoneHash
+    )
+
+    const msgHash = contractKit.web3.utils.soliditySha3({ type: 'address', value: walletAddress })
+    const context = newTransactionContext(TAG, 'Withdraw from escrow')
+    // TODO: Batch the tranasctions and submit them together via `executeTransactions`
+    // method on an instance of the MTW then submitting like usual
+    const withdrawTxSuccess: boolean[] = []
+    for (let i = 0; i < escrowPaymentIds.length; i += 1) {
+      const withdrawTx: CeloTransactionObject<boolean> = yield formEscrowWithdrawTxWithNoCode(
+        contractKit,
+        escrowWrapper,
+        escrowPaymentIds[i],
+        pepper,
+        msgHash,
+        i
+      )
+
+      try {
+        if (!komenciActive) {
+          yield call(sendTransaction, withdrawTx.txo, walletAddress, context)
+        } else {
+          const withdrawTxResult: Result<TransactionReceipt, FetchError | TxError> = yield call(
```

Comment:
We should probably ensure we have an active session here before we try this call? If a session is not open, can open a new session here?
One case I can imagine here is that the user closed the application after verification (e.g. left their phone on the table while they wait) and came back to it 15 minutes later when their session with Komenci is no longer valid.

Reply:
That would be a more complete approach but I'm reluctant to do that until we've built a way for users to initiate a redemption outside of the automatic one done during the onboarding flow. While the edge case you pointed out is possible, in practice that would be very rare because Komenci has very lax session expiration criteria (I think it's the first of one week passing or >20 metatx's subsidized)